pkg/remotecmd/kubeexec.go

```diff
@@ -318,40 +318,82 @@ func (k *kubeExecProcessHandler) getProcessInfoFromPid(
+		klog.V(7).Infof("stderr: %s\n", strings.Join(stderr, "\n"))
+		return nil, err
+	}
+
+	allProcesses := make(map[int][]int)
+	for _, line := range stdout {
+		var pid int
+		_, err = fmt.Sscanf(line, "%d ", &pid)
+		if err != nil {
+			return nil, err
+		}
+
+		// Last index of the last ")" character to unambiguously parse the content.
+		// See https://unix.stackexchange.com/questions/558239/way-to-unambiguously-parse-proc-pid-stat-given-arbitrary-contents-of-name-fi
+		i := strings.LastIndex(line, ")")
+		if i < 0 {
+			continue
+		}
+
+		// At this point, "i" is the index of the last ")" character, and we have an additional space before the process state character, hence the "i+2".
+		// For example:
+		// 87 (main) S 0 81 81 0 -1 ...
+		var state byte
+		var ppid int
+		_, err = fmt.Sscanf(line[i+2:], "%c %d", &state, &ppid)
+		if err != nil {
+			return nil, err
+		}
+
+		allProcesses[ppid] = append(allProcesses[ppid], pid)
+	}
+
+	return allProcesses, nil
+}
+
 // getProcessChildren returns all the children (either direct or indirect) of the specified process in the given container.
-// It works by reading the /proc/<pid>/task/<tid>/children file, which is a space-separated list of children,
-// and then recursively finds the children for each child.
+// It works by reading the /proc/<pid>/stat files, giving PPID for each PID.
 // The overall result is an ordered list of children PIDs obtained via a recursive post-order traversal algorithm.
 func (k *kubeExecProcessHandler) getProcessChildren(pid int, podName string, containerName string) ([]int, error) {
 	if pid <= 0 {
 		return nil, fmt.Errorf("invalid pid: %d", pid)
 	}
 
-	stdout, _, err := k.execClient.ExecuteCommand(
-		[]string{ShellExecutable, "-c", fmt.Sprintf("cat /proc/%[1]d/task/%[1]d/children || true", pid)},
-		podName, containerName, false, nil, nil)
+	allProcesses, err := k.getAllProcesses(podName, containerName)
 	if err != nil {
 		return nil, err
 	}
 
-	var children []int
-	for _, line := range stdout {
-		l := strings.Split(strings.TrimSpace(line), " ")
-		for _, p := range l {
-			c, err := strconv.Atoi(p)
-			if err != nil {
-				return children, err
-			}
-			childChildren, err := k.getProcessChildren(c, podName, containerName)
-			if err != nil {
-				return children, err
-			}
-			children = append(children, childChildren...)
-			children = append(children, c)
+	return k.getProcessChildrenRec(allProcesses, pid)
+}
+
+// getProcessChildrenRec return an ordered list of children PIDs of the given pid,
+// using a recursive post-order traversal algorithm so that the returned list can start with the deepest children processes.
+// The allProcesses map is the parent-children process hierarchy of all processes, where the key is a parent PID,
+// and the value is a list of all its direct children. See the getAllProcesses method.
+func (k *kubeExecProcessHandler) getProcessChildrenRec(allProcesses map[int][]int, pid int) ([]int, error) {
```

Comment:
nit:
I was wondering if this can be defined as an internal function within `k.getProcessChildren`.

Reply:
Since this is a recursive function, I am not sure if it would be possible to declare it within `k.getProcessChildren`. Well, I need to check, but maybe by separating the variable declaration and assignment.

Reply:
Okay, indeed, as I suspected, it won't compile if I declare it right in `k.getProcessChildren`:

```go
func (k *kubeExecProcessHandler) getProcessChildren(pid int, podName string, containerName string) ([]int, error) {
	//...
	getProcessChildrenRec := func(allProcesses map[int][]int, pid int) []int {
		// ...
		// This won't compile
		var allChildren []int
		for _, child := range allProcesses[pid] {
			allChildren = append(allChildren, getProcessChildrenRec(allProcesses, child)...)
		}
		// ...
	}
	return getProcessChildrenRec(allProcesses, pid), nil
}
```

But separating declaration and assignment works, but seems a bit redundant to me:

```go
func (k *kubeExecProcessHandler) getProcessChildren(pid int, podName string, containerName string) ([]int, error) {
	//...
	var getProcessChildrenRec func(allProcesses map[int][]int, pid int) []int
	getProcessChildrenRec := func(allProcesses map[int][]int, pid int) []int {
		// ...
		var allChildren []int
		for _, child := range allProcesses[pid] {
			allChildren = append(allChildren, getProcessChildrenRec(allProcesses, child)...)
		}
		// ...
	}
	return getProcessChildrenRec(allProcesses, pid), nil
}
```

I'll keep it as it is, but will change it as a function (as it does not need the receiver).

Reply:
Actually, it could be acceptable by simplifying the closure parameters. Let's make the change then :-)

Reply:
Updated in [`c234978` (#6690)](https://github.com/redhat-developer/odo/pull/6690/commits/c234978fb19f5cee3be1c4c9577432d31e723517)